src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +257,300 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
 
 	return true;
 }
+
+
+/*
+ * streamBufferedXID reads a transaction buffer file and streams it into the
+ * target database.
+ */
+static bool
+streamBufferedXID(ReplayStreamCtx *replayCtx)
+{
+	/* Open replaybuffer in read-only mode to use it as input stream */
+	FILE *txnBufferForRead = fopen_read_only(replayCtx->currentTxnFileName);
+
+	if (txnBufferForRead == NULL)
+	{
+		log_error("Failed to open transaction buffer file \"%s\" in "
+				  "readonly mode: %m",
+				  replayCtx->currentTxnFileName);
+		return false;
+	}
+
+	ReadFromStreamContext readerContext = {
+		.callback = stream_replay_line,
+		.ctx = replayCtx
+	};
+
+	if (!read_from_stream(txnBufferForRead, &readerContext))
+	{
+		log_error("Failed to read from replay buffer, "
+				  "see above for details");
+		return false;
+	}
+
+	fclose(txnBufferForRead);
+
+	return true;
+}
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file when we encounter
+ * a COMMIT without a valid txnCommitLSN.
+ *
+ * This ensures that the apply process doesn't block the generation of the
+ * transaction metadata file created by the transform process on arrival of
+ * a COMMIT message.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	LogicalMessageMetadata metadata = { 0 };
+
+	if (!parseSQLAction((char *) line, &metadata))
+	{
+		/* errors have already been logged */
+		return false;
+	}
+
+	/* We don't use this metadata in this function */
+	free(metadata.jsonBuffer);
+
+
+	if (metadata.action == STREAM_ACTION_BEGIN)
+	{
+		/*
+		 * Enable transaction buffering only when the BEGIN doesn't have
+		 * valid commit LSN, otherwise we can skip buffering and stream
+		 * the transaction directly.
+		 */
+		replayCtx->skipTxnBuffering = metadata.txnCommitLSN != InvalidXLogRecPtr;
+		if (replayCtx->skipTxnBuffering)
+		{
+			return stream_replay_line(ctx, line, stop);
+		}
+
+		if (replayCtx->isTxnBuffering)
+		{
+			/*
+			 * When the follow switches from prefetch to replay mode, we
+			 * call stream_transform_stream which might stream the partially
+			 * written transaction created during the prefetch mode.
+			 *
+			 * For example, lets consider the partially written C.sql file,
+			 *
+			 * BEGIN -- {"xid": 1000, "commitLSN": "0/1234"};
+			 * INSERT INTO C VALUES (1);
+			 * INSERT INTO C VALUES (2);
+			 * KEEPALIVE;
+			 *
+			 * After switching to the replay mode, logical decoding will resume
+			 * from consistent point which again starts with a valid transcation
+			 * block.
```

Comment:
Do we have the LSN value for that consistent point available in the context here? If yes, I believe we should be using it to make sure we are skipping transactions as intended. If no, then we should add to the comment why not, so that another PR later can go and fix the situation if we feel like it should be fixed.

Reply:
@dimitri  This is a great point !  I think the current ignore logic implemented in this PR might cause inconsistency where it might ignore the partial transaction, but applies the next transaction because it has valid BEGIN/COMMIT block. 

I need to have a another deep look now.

Reply:
@dimitri I was wrong in my previous comment. 

While switching to replay mode, follow ensures that the pending LSNs in transform queue has been [processed and applied to the target](https://github.com/dimitri/pgcopydb/blob/672ad1228278f22cd450b214afc3e0092ac47a15/src/bin/pgcopydb/follow.c#L429-L489).  This ensures everything except the last the partial transaction has been successfully committed into the target. After switching to replay mode, [stream_transform_resume](https://github.com/dimitri/pgcopydb/blob/672ad1228278f22cd450b214afc3e0092ac47a15/src/bin/pgcopydb/ld_transform.c#L223-L253) again streams the last partial file(which is already applied by the previous step) to the replay process which leads to this scenario. I think streaming last partial .sql from [stream_transform_resume](https://github.com/dimitri/pgcopydb/blob/672ad1228278f22cd450b214afc3e0092ac47a15/src/bin/pgcopydb/ld_transform.c#L223-L253) is redundant and shall be removed and this shouldn't cause any issue. WDYT?

Reply:
I'm wondering are these PRs https://github.com/dimitri/pgcopydb/pull/348, https://github.com/dimitri/pgcopydb/pull/277 trying to do the same in different ways?

Reply:
IIRC the transform parts need to parse/load the latest file to build up the context in case we stopped in the middle of a transaction (we then need to have the start of the transaction in-memory); but I agree we don't need to send that to the replay process, which is only going to skip again.

That said the skip logic is sound and I wonder why in your PR we can't benefit from it?